src/sagemaker/image.py

```diff
@@ -0,0 +1,446 @@
+            shutil.copytree(data_dir, os.path.join(self.container_root, host, 'input', 'data'))
+
+        compose_data = self._generate_compose_file('train', additional_volumes=volumes)
+        compose_command = self._compose()
+        _stream_output(compose_command)
+        _cleanup()
+
+        s3_model_artifacts = self.retrieve_model_artifacts(compose_data)
+        return s3_model_artifacts
+
+    def serve(self, primary_container):
+        """Host a local endpoint using docker-compose.
+        Args:
+            primary_container (dict): dictionary containing the container runtime settings
+                for serving. Expected keys:
+                - 'ModelDataUrl' pointing to either an s3 location or a local file
+                - 'Environment' a dictionary of environment variables to be passed to the hosting container.
+
+        """
+        logger.info("serving")
+
+        self.container_root = _create_tmp_folder()
+        logger.info('creating hosting dir in {}'.format(self.container_root))
+
+        model_dir = primary_container['ModelDataUrl']
+        if not model_dir.lower().startswith("s3://"):
+            for h in self.hosts:
+                host_dir = os.path.join(self.container_root, h)
+                os.makedirs(host_dir)
+                shutil.copytree(model_dir, os.path.join(self.container_root, h, 'model'))
+
+        env_vars = ['{}={}'.format(k, v) for k, v in primary_container['Environment'].items()]
+
+        self._generate_compose_file('serve', additional_env_vars=env_vars)
+        compose_command = self._compose()
+        self.container = _Container(self.container_root, compose_command)
+        self.container.up()
+
+    def stop_serving(self):
+        """Stop the serving container.
+
+        The serving container runs in async mode to allow the SDK to do other tasks.
+        """
+        if self.container:
+            self.container.down()
+
+    def retrieve_model_artifacts(self, compose_data):
+        """Get the model artifacts from all the container nodes.
+
+        Used after training completes to gather the data from all the individual containers. As the
+        official SageMaker Training Service, it will override duplicate files if multiple containers have
+        the same file names.
+
+        Args:
+            compose_data(dict): Docker-Compose configuration in dictionary format.
+
+        Returns: Local path to the collected model artifacts.
+
+        """
+        # Grab the model artifacts from all the Nodes.
+        s3_model_artifacts = os.path.join(self.container_root, 's3_model_artifacts')
+        os.mkdir(s3_model_artifacts)
+
+        for host in self.hosts:
+            volumes = compose_data['services'][str(host)]['volumes']
+
+            for volume in volumes:
+                container_dir, host_dir = volume.split(':')
+                if host_dir == '/opt/ml/model':
+                    self._recursive_copy(container_dir, s3_model_artifacts)
```

Comment:
because shutil.copytree assumes that the target directory does not exist. When doing distributed training, the target dir will exist after getting the data out of the first container.